### ISSUE TITLE
Segmentation fault on BT destruction.

### ISSUE THREAD

**Issue by Krzo99** [BehaviorTree/BehaviorTree.CPP#816]
I came across a problem that only happens around 5-10% of the time.

It appers that a segmentation fault is thrown on destruction of BT::Tree object.

My setup is as follows:

First, I spin up a seperate thread, that will run Behavior Trees:  `std::thread{ std::bind(&AutonomousRobotControl::StartbehaviorTreeThread, this) }.detach();`

In the thread, I construct all objects:
```
BT::BehaviorTreeFactory factory;

factory.registerNodeType<BT_Node::Navigate>("Navigate", navigator_);
factory.registerNodeType<BT_Node::UpdateGoal>("UpdateGoal", navigator_);
....
factory.registerSimpleCondition("isLocalized", std::bind(BT_Condition::isLocalized, rtabmap_manager_));
factory.registerSimpleCondition("isColdStart", std::bind(BT_Condition::isColdStart, rtabmap_manager_));
....

auto tree = factory.createTreeFromFile(behavior_tree_xml_);
BT::Blackboard::Ptr blackboard = tree.rootBlackboard();

blackboard->set("current_action", RobotState::IDLE);
....

std::unique_ptr<BT::StdCoutLogger> logger;
if (debug_mode_)
    logger = std::make_unique<BT::StdCoutLogger>(tree);

 BT::Groot2Publisher groot2_telemetry_publisher(
    tree, 
    1667 + parent_->get_parameter("robot_id").as_int() * 2);
```

Then, I start the behavior loop:

```
while(rclcpp::ok()) {
            if (is_stopping_) {
                is_running_ = false;
                break;
            }
            tree.tickOnce();

        rate.sleep();
    }

```

Behavior tree execution is stopped, once is_stopping_ is set from the main thread. 
(There is a mutex to synchronize access, I just didn't include it in this code snip)

The errors I'm getting is always a bit different every time, however what seems to be common denominator, in that they are all based around (and they also happen) when behavior tree destructs, as seen in the logs.

I'm not sure what the problem is. Would it be possible that I'm stopping the behavior tree execution wrongly?
Thanks for your help!

Down bellow, I provide 2 Callgrind traces, and one gdb.

Callgrind 1:
```
[valgrind-7] ==139311== Process terminating with default action of signal 11 (SIGSEGV)
[valgrind-7] ==139311==  Access not within mapped region at address 0x1C00002D8
[valgrind-7] ==139311==    at 0x5376857: _int_free (malloc.c:4590)
[valgrind-7] ==139311==    by 0x5379452: free (malloc.c:3391)
[valgrind-7] ==139311==    by 0x563AD46: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, BT::TreeNodeManifest>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, BT::TreeNodeManifest> >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::clear() (new_allocator.h:145)
[valgrind-7] ==139311==    by 0x5625E37: BT::BehaviorTreeFactory::~BehaviorTreeFactory() (hashtable.h:1532)
[valgrind-7] ==139311==    by 0x4875B14: terrascout_control::AutonomousRobotControl::StartbehaviorTreeThread() (autonomous_robot_control.cpp:237)
[valgrind-7] ==139311==    by 0x5164252: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30)
[valgrind-7] ==139311==    by 0x5368AC2: start_thread (pthread_create.c:442)
[valgrind-7] ==139311==    by 0x53F9A03: clone (clone.S:100)
[valgrind-7] ==139311==  If you believe this happened as a result of a stack
[valgrind-7] ==139311==  overflow in your program's main thread (unlikely but
[valgrind-7] ==139311==  possible), you can try to increase the size of the
[valgrind-7] ==139311==  main thread stack using the --main-stacksize= flag.
[valgrind-7] ==139311==  The main thread stack size used in this run was 8388608.
[valgrind-7] ==139311== 
[valgrind-7] --139311-- Start dumping at BB 0 (Prg.Term.)...
[valgrind-7] --139311-- Dump to /home/domen/callgrind.out.139311
[valgrind-7] --139311-- Dumping done.
[valgrind-7] ==139311== Events    : Ir
[valgrind-7] ==139311== Collected : 0
[valgrind-7] ==139311== 
[valgrind-7] ==139311== I   refs:      0
```

Callgrind 2:

```
[valgrind-7] scout_control_unit: pthread_mutex_lock.c:94: ___pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
[valgrind-7] ==70790== 
[valgrind-7] ==70790== Process terminating with default action of signal 6 (SIGABRT)
[valgrind-7] ==70790==    at 0x53739FC: pthread_kill@@GLIBC_2.34 (pthread_kill.c:44)
[valgrind-7] ==70790==    by 0x531F475: raise (raise.c:26)
[valgrind-7] ==70790==    by 0x53057F2: abort (abort.c:79)
[valgrind-7] ==70790==    by 0x530571A: __assert_fail_base.cold (assert.c:92)
[valgrind-7] ==70790==    by 0x5316E95: __assert_fail (assert.c:101)
[valgrind-7] ==70790==    by 0x53750CF: pthread_mutex_lock@@GLIBC_2.2.5 (pthread_mutex_lock.c:94)
[valgrind-7] ==70790==    by 0x566057A: BT::TreeNode::status() const (gthr-default.h:749)
[valgrind-7] ==70790==    by 0x5660EFB: BT::ControlNode::resetChildren() (control_node.cpp:42)
[valgrind-7] ==70790==    by 0x5660F2C: BT::ControlNode::halt() (control_node.cpp:34)
[valgrind-7] ==70790==    by 0x5660D22: BT::TreeNode::haltNode() (tree_node.cpp:141)
[valgrind-7] ==70790==    by 0x5634B58: BT::applyRecursiveVisitor(BT::TreeNode*, std::function<void (BT::TreeNode*)> const&) (std_function.h:590)
[valgrind-7] ==70790==    by 0x5634B9C: BT::applyRecursiveVisitor(BT::TreeNode*, std::function<void (BT::TreeNode*)> const&) (behavior_tree.cpp:54)
[valgrind-7] ==70790== 
[valgrind-7] --70790-- Start dumping at BB 0 (Prg.Term.)...
[valgrind-7] --70790-- Dump to /home/domen/Documents/TerraScout/core/callgrind.out.70790
[valgrind-7] --70790-- Dumping done.
[valgrind-7] ==70790== Events    : Ir
[valgrind-7] ==70790== Collected : 0
[valgrind-7] ==70790== 
[valgrind-7] ==70790== I   refs:      0
```

gdb stack trace:

```
[gdb-7] [Switching to Thread 0x7fff137fe000 (LWP 131226)]
[gdb-7] 0x0000000000000100 in ?? ()
[gdb-7] #0  0x0000000000000100 in  ()
[gdb-7] #1  0x00007ffff736fad9 in std::_Function_base::~_Function_base()
[gdb-7]     (this=0x7fff18038a70, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/std_function.h:244
[gdb-7] #2  std::function<BT::Any (BT::Ast::Environment&)>::~function()
[gdb-7]     (this=0x7fff18038a70, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/std_function.h:334
[gdb-7] #3  std::array<std::function<BT::Any (BT::Ast::Environment&)>, 4ul>::~array()
[gdb-7]     (this=0x7fff18038a30, this=<optimized out>) at /usr/include/c++/11/array:95
[gdb-7] #4  BT::TreeNode::PImpl::~PImpl() (this=0x7fff18038760, this=<optimized out>)
[gdb-7]     at ./src/tree_node.cpp:21
[gdb-7] #5  std::default_delete<BT::TreeNode::PImpl>::operator()(BT::TreeNode::PImpl*) const (this=<optimized out>, __ptr=0x7fff18038760)
[gdb-7]     at /usr/include/c++/11/bits/unique_ptr.h:85
[gdb-7] #6  std::default_delete<BT::TreeNode::PImpl>::operator()(BT::TreeNode::PImpl*) const (__ptr=0x7fff18038760, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/unique_ptr.h:79
[gdb-7] #7  std::unique_ptr<BT::TreeNode::PImpl, std::default_delete<BT::TreeNode::PImpl> >::~unique_ptr() (this=<optimized out>, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/unique_ptr.h:361
[gdb-7] #8  BT::TreeNode::~TreeNode() (this=<optimized out>, this=<optimized out>)
[gdb-7]     at ./src/tree_node.cpp:71
[gdb-7] #9  0x00007ffff73617c3 in BT::ControlNode::~ControlNode()
[gdb-7]     (this=0x7fff18005ac0, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/allocator.h:174
[gdb-7] #10 BT::FallbackNode::~FallbackNode()
[gdb-7]     (this=0x7fff18005ac0, this=<optimized out>)
[gdb-7]     at ./include/behaviortree_cpp/controls/fallback_node.h:37
[gdb-7] #11 BT::FallbackNode::~FallbackNode()
[gdb-7]     (this=0x7fff18005ac0, this=<optimized out>)
[gdb-7]     at ./include/behaviortree_cpp/controls/fallback_node.h:37
[gdb-7] #12 0x00007ffff7385d5c in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x7fff18038ac0)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:168
[gdb-7] #13 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count()
[gdb-7]     (this=0x7fff180403c8, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:705
[gdb-7] #14 std::__shared_ptr<BT::TreeNode, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7fff180403c0, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:1154
[gdb-7] #15 std::shared_ptr<BT::TreeNode>::~shared_ptr()
[gdb-7]     (this=0x7fff180403c0, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr.h:122
[gdb-7] #16 std::destroy_at<std::shared_ptr<BT::TreeNode> >(std::shared_ptr<BT::TreeNode>*) (__location=0x7fff180403c0) at /usr/include/c++/11/bits/stl_construct.h:88
[gdb-7] #17 std::_Destroy<std::shared_ptr<BT::TreeNode> >(std::shared_ptr<BT::TreeNode>*) (__pointer=0x7fff180403c0) at /usr/include/c++/11/bits/stl_construct.h:149
[gdb-7] #18 std::_Destroy_aux<false>::__destroy<std::shared_ptr<BT::TreeNode>*>(std::shared_ptr<BT::TreeNode>*, std::shared_ptr<BT::TreeNode>*)
[gdb-7]     (__last=0x7fff18040420, __first=0x7fff180403c0)
[gdb-7]     at /usr/include/c++/11/bits/stl_construct.h:163
[gdb-7] #19 std::_Destroy<std::shared_ptr<BT::TreeNode>*>(std::shared_ptr<BT::TreeNode>*, std::shared_ptr<BT::TreeNode>*)
[gdb-7]     (__last=0x7fff18040420, __first=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/stl_construct.h:196
[gdb-7] #20 std::_Destroy<std::shared_ptr<BT::TreeNode>*, std::shared_ptr<BT::TreeNode> >(std::shared_ptr<BT::TreeNode>*, std::shared_ptr<BT::TreeNode>*, std::allocator<std::shared_ptr<BT::TreeNode> >&)
[gdb-7]     (__last=0x7fff18040420, __first=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/alloc_traits.h:848
[gdb-7] #21 std::vector<std::shared_ptr<BT::TreeNode>, std::allocator<std::shared_ptr<BT::TreeNode> > >::~vector() (this=0x7fff18020430, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/stl_vector.h:680
[gdb-7] #22 BT::Tree::Subtree::~Subtree() (this=0x7fff18020430, this=<optimized out>)
[gdb-7]     at ./include/behaviortree_cpp/bt_factory.h:98
[gdb-7] #23 std::destroy_at<BT::Tree::Subtree>(BT::Tree::Subtree*)
[gdb-7]     (__location=0x7fff18020430) at /usr/include/c++/11/bits/stl_construct.h:88
[gdb-7] #24 std::allocator_traits<std::allocator<BT::Tree::Subtree> >::destroy<BT::Tree::Subtree>(std::allocator<BT::Tree::Subtree>&, BT::Tree::Subtree*)
[gdb-7]     (__p=0x7fff18020430, __a=...)
[gdb-7]     at /usr/include/c++/11/bits/alloc_traits.h:537
[gdb-7] #25 std::_Sp_counted_ptr_inplace<BT::Tree::Subtree, std::allocator<BT::Tree::Subtree>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() (this=0x7fff18020420)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:528
[gdb-7] #26 0x00007ffff73486f4 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x7fff18020420)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:168
[gdb-7] #27 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count()
[gdb-7]     (this=0x7fff1803a078, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:705
[gdb-7] #28 std::__shared_ptr<BT::Tree::Subtree, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7fff1803a070, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr_base.h:1154
[gdb-7] #29 std::shared_ptr<BT::Tree::Subtree>::~shared_ptr()
[gdb-7]     (this=0x7fff1803a070, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/shared_ptr.h:122
[gdb-7] #30 std::destroy_at<std::shared_ptr<BT::Tree::Subtree> >(std::shared_ptr<BT::Tree::Subtree>*) (__location=0x7fff1803a070)
[gdb-7]     at /usr/include/c++/11/bits/stl_construct.h:88
[gdb-7] #31 std::_Destroy<std::shared_ptr<BT::Tree::Subtree> >(std::shared_ptr<BT::Tree::Subtree>*) (__pointer=0x7fff1803a070)
[gdb-7]     at /usr/include/c++/11/bits/stl_construct.h:149
[gdb-7] #32 std::_Destroy_aux<false>::__destroy<std::shared_ptr<BT::Tree::Subtree>*>(std::shared_ptr<BT::Tree::Subtree>*, std::shared_ptr<BT::Tree::Subtree>*)
[gdb-7]     (__last=0x7fff1803a090, __first=0x7fff1803a070)
[gdb-7]     at /usr/include/c++/11/bits/stl_construct.h:163
[gdb-7] #33 std::_Destroy<std::shared_ptr<BT::Tree::Subtree>*>(std::shared_ptr<BT::Tree::Subtree>*, std::shared_ptr<BT::Tree::Subtree>*)
[gdb-7]     (__last=0x7fff1803a090, __first=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/stl_construct.h:196
[gdb-7] #34 std::_Destroy<std::shared_ptr<BT::Tree::Subtree>*, std::shared_ptr<BT::Tree::Subtree> >(std::shared_ptr<BT::Tree::Subtree>*, std::shared_ptr<BT::Tree::Subtree>*, std::allocator<std::shared_ptr<BT::Tree::Subtree> >&)
[gdb-7]     (__last=0x7fff1803a090, __first=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/alloc_traits.h:848
[gdb-7] #35 std::vector<std::shared_ptr<BT::Tree::Subtree>, std::allocator<std::shared_ptr<BT::Tree::Subtree> > >::~vector()
[gdb-7]     (this=0x7fff137c3660, this=<optimized out>)
[gdb-7]     at /usr/include/c++/11/bits/stl_vector.h:680
[gdb-7] #36 BT::Tree::~Tree() (this=0x7fff137c3660, this=<optimized out>)
[gdb-7]     at ./src/bt_factory.cpp:573
[gdb-7] #37 0x00007ffff7f82b0b in terrascout_control::AutonomousRobotControl::StartbehaviorTreeThread() (this=0x555556850750)
[gdb-7]     at /home/domen/Documents/TerraScout/core/src/terrascout/terrascout_control/src/autonomous_robot_control.cpp:237
[gdb-7] #38 0x00007ffff74dc253 in  () at /lib/x86_64-linux-gnu/libstdc++.so.6
[gdb-7] #39 0x00007ffff7094ac3 in start_thread (arg=<optimized out>)
[gdb-7]     at ./nptl/pthread_create.c:442
[gdb-7] #40 0x00007ffff7126850 in clone3 ()
[gdb-7]     at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

```

**Comment by facontidavide** [comment]
Did you upgrade recently to version 4.6?

If that is the case, it might be a problem of ABI compatibility. Try recompiling from scratch everything, including the plugins.

Otherwise, we can try using **address sanitizer** to get more information.

The info you shared is not giving me any hypothesis, for the time being.

**Comment by Krzo99** [comment]
Ok, so it appears that the problem was the fact that I was manually .reset() a shared_ptr with a custom class that was used in some Nodes in the tree, at the end of thread() function. 
I'm guessing there was a possibility for a race condition here. I just moved the .reset() call to after the thread joined.
I also changed a detachable thread to a joinable one, as that feels more accurate.

Currently, after 2h of testing, there hasn't been another error like this, so I'm guessing I solved it :)
